jena-arq/src/main/java/org/apache/jena/sparql/graph/PrefixMappingBase.java

```diff
@@ -0,0 +1,277 @@
+     * 
+     * PrefixMappingMem does.
+     * Database backed ones typically don't.
+     */
+
+    protected PrefixMappingBase() {}
+    
+    // The storage operations of an implementation.
+    
+    /** Add prefix */
+    abstract protected void add(String prefix, String uri);
+    
+    /** Remove prefix. */
+    abstract protected void remove(String prefix);
+    
+    /** Clear all mappings */
+    abstract protected void clear();
+
+    abstract protected boolean isEmpty();
+    
+    abstract protected int size();
+
+    /** Return the URI that the prefix maps to. */ 
+    abstract protected String prefixToUri(String prefix);
+    
+    /** Return a prefix that maps to the URI.
+     * There may be several; the answer is any one of them. 
+     */ 
+    abstract protected String uriToPrefix(String uri);
+    
+    /** Return as a map. This map is only used within this class.
+     * It can be as efficient as possible.
+     * It will not be modifed.
+     * It wil not be retuned to a caller of {@code PrefixMappingBase}.
+     */
+    abstract protected Map<String, String> asMap();
+
+    /** 
+     * Return as a map. The map return is not connected to the prefix mapping implementation,
+     * does not reflect subsequent prefix mapping changes.
+     */
+    abstract protected Map<String, String> asMapCopy();
+
+    /** Apply the {@link BiConsumer} to each (prefix, uri) pair. */
+    abstract protected void apply(BiConsumer<String, String> action);
+    
+    /**
+     * This part of the subclass API and may be overriden if an implementation can do
+     * better This general implementation is based on asMap() which may be a copy or may
+     * be a view of the storage directly.
+     */
+    protected Optional<Entry<String, String>> findMapping( String uri, boolean partial ) {
+        return asMap().entrySet().stream().sequential().filter(e->{
+            String ss = e.getValue();
+            if (uri.startsWith( ss ) && (partial || ss.length() == uri.length())) 
+                return true;
+            return false;
+        }).findFirst();
+    }
+
+    @Override
+    public PrefixMapping setNsPrefix(String prefix, String uri) {
+        checkLegal(prefix); 
+        add(prefix, uri);
+        return this;
+    }
+    
+    @Override
+    public PrefixMapping removeNsPrefix(String prefix) {
+        remove(prefix);
+        return this;
+    }
+    
+    @Override
+    public PrefixMapping clearNsPrefixMap() {
+        clear();
+        return this;
+    }
+
+    /**
+     * Checks that a prefix is "legal" - it must be a valid XML NCName or "". XML rules
+     * for RDF/XML output.
+     * <p>
+     * This is a recurring user question - why does {@code Resource.getNamespace},
+     * {@code Resource.getLocalname} not abbreviate when it is legal Turtle.
+     * <p>
+     * Answer - legacy for RDF/XML.
+     * <p>
+     * See also {@link #qnameFor}.
+     */
+    private void checkLegal(String prefix) {
```

Comment:
Could this be `public static` so this useful comment is included in Javadoc?

Reply:
Done, with named changed to `checkLegalPrefix`.  Typos fixed as well.